### ISSUE TITLE
Erk's suggestions for NPC Scripting Language - first steps

### ISSUE THREAD

**Issue by I-am-Erk** [CleverRaven/Cataclysm-DDA#30936]
Scripts are a way to manage NPC behaviour, to get NPCs to take specific actions in response to events. They ate complementary to AI and necessary for immersive storytelling. While AI allows emergent behaviour, scripting allows situation specific behaviours. A good scripting language will also work closely with AI so that the AI occasionally calls particular scripts on its own. To make all this work, we need a way to script NPC behaviours and interactions with their environment.

This outlines how I see a scripting language works. There are three basic pieces: the script definition, the script call, and the script trigger.

### Definition and Structure
`npc_script` is a type of JSON entry that defines discrete NPC actions as elements within the entry. Eventually it can run once or loop, and events can be scheduled or occur immediately. For first pass we can probably handle events running once immediately.

I see four fairly straightforward first pass scripts: descriptive text, movement, start conversation, and add effect.

 Any one of these could be done first.  Of course a single script could call any number of these in combination.
 
This Issue will only describe a good first pass of scripts. I haven't got into many of the more niche actions I'd love to see (eg interacting with items and furniture on the map, looping scripts that set up an NPC's daily routine, etc) because for now, this will be plenty.
 
#### Some terms all scripts need
* `"wait": "time unit"` causes the NPC to wait for the specified duration in turns or s/m/h before executing the next command in the script
* `"repeat": int` repeats this element of the script a number of times equal to int.

#### NPC descriptive text scripts
These are very simple scripts that just have the NPC output a bit of descriptive text at a specified volume. This could be used for flavourful actions, or to have the NPC say something out loud. Aside from potentially making a noise, it doesn't have any direct effect on the game.

Most basic example:
```
{ 
 "type": "npc_script",
 "id": "refugee_draco_tune",
 "action": {
   "text": "Draco tunes his guitar.",
   "volume": 5
  }
}
```
This happens once, and could be specified eg. at a random interval between 1-2 hours (see triggers) so that every so often, Draco tunes his guitar.
```
{ 
 "type": "npc_script",
 "id": "beggar_luo_complain",
 "action": {
   "text": [ "Luo grumbles under her breath.", "Luo mutters \"Wish I had somewhere comfortable to sit.\"", "Luo says, \"<swear!>!  Did you just fart, Dave?\" ]
   "volume": 5
  }
}
```
This would be another random interval script, but has an array of possible options that would be chosen at random, and uses a text snippet.
 
 #### NPC movement scripts
 These would allow an NPC from move to point A to point B, using normal pathfinding along the way.  Any NPC that has a movement script should have an `origin`, the place it will go back to after the script ends. If not defined, the `origin` defaults to the tripoint location the NPC started on.  The destination is defined either by relative x,y,z coordinates from the origin, or by x,y locations on the nearest om_special of a specific type.
 
 It may be useful to have a way to move towards or away from the activator of the script as well.
 
 Movements can specify speed: `run`, `walk`, and `crouch` would move at maximum speed in those stances. Optionally we could add `meander`, which would make the NPC walk to the location specified at a random slow rate, something like 1/2 their normal movement speed.
 
 Examples:
 ```
{ 
 "type": "npc_script",
 "id": "refugee_jenny_walk_to_bathroom",
 "action": [ "move": {
   "x": 6,
   "y": 15,
   "rotate_with_map": true,
   "speed": "walk"
  },
  "wait": "5 m"
  ]
 }
 ```
 This will cause the NPC to move to a location 5 tiles right and 15 tiles down from their starting location.  Because it rotates with the map, if the refugee center spawned rotated, the x,y coordinates will change so that she doesn't randomly walk outside instead of to the toilet. When she gets there, she'll run the next action in the list, which is 'wait' for 5 minutes. When she's done, she'll return to the origin.
 
 If I wanted to, I could have her wait in the toilet, then walk to the sink and use a descriptive text to have her wash her hands. Wash your damn hands, Jenny, that's gross.

 ```
{ 
 "type": "npc_script",
 "id": "guard_intercept",
 "action": { "move_to_trigger": 5
   "speed": "run" }
 }
 ```
In this example, the guard will run up to 5 squares, towards the character that triggered the movement. This might be used to have a guard follow the player around and stop them from stealing, or to have a guard run up and tell you you're in a restricted area.

  ```
{ 
 "type": "npc_script",
 "id": "arsonist_travel_home",
 "action": [
   "move_to_om": { 
     "dest": "fort_evac",
     "x": 8,
     "y": 12,
     "rotate_with_map": true,
     "reset_origin": true,
     "speed": "walk"
    } 
  ]
 }
 ```
 For longer distance moves, or moves to a totally different OM, `move_to_om` would be the thing to do. This example will send the NPC to the nearest om_special named "fort evac", where she will head to x8,y12 on that om_special and declare it her new origin tile for the purpose of investigating sounds, running relative movement scripts, and things.
 
 Presumably this will cause problems with reality bubbles and things. I do not think we need to fix that problem just yet; however, the best way I can think of to manage this would be for the game to estimate an overland travel time, and if the NPC leaves the reality bubble they could be loaded into the buffer and placed at the specified location the next time it is loaded provided the travel time has passed.
 
 #### Start Conversation
 This is a super easy one: When this script is activated, the NPC starts a conversation with the player. Obviously this will need to be associated with some kind of trigger - see script calls below. This is something we sorely need, to cause guards and similar NPCs to challenge you when you try to enter their protected area.
 
 #### Add effect
 This script would add an effect as per effects that can be added in conversation, using the same syntax.  This would allow countless things to happen as the result of a script, such as having an NPC start or stop following the player when they activate a certain trigger, or activating a variable so that the NPC comments in conversation that the player did something. 

### Script Calls
Scripts are called from two places, the `npc` definition or in effects (and thus from dialogue or missions)

#### NPC definition calls
In the NPC definition, each script would have a trigger to run. The most common triggers would be random intervals, or related to spotting the player, a monster, a character of another faction, stuff like that. See Triggers, below. Here's what I picture a script call in an NPC definition looking like.

```
"scripts": 
[
  { "script": "guard_noentry", "trigger": "nonfaction_approach", "distance": 3 },
  { "script": "guard_attack", "trigger": "nonfaction_approach", "distance": 1 },
  { "script": "guard_greeting", "trigger": "faction_approach", "distance": 3, "prob": 50 }
]
```

This example is a stationary guard NPC that will stop a nonfaction character from passing if they get too close, by running a script called guard_noentry (which presumably has them say something like "you can't go there"). If a faction character approaches, the guard just says a greeting half the time, or does nothing the other half. This guard would only run its script if the character was within 3 tiles of the guard.  If the player moves within 1 square of the guard, the guard becomes hostile (through a script that runs a dialogue effect).

#### Effect calls
Scripts run as an effect, the sort called from dialogues or missions (or, if I get my way, from other scripts: this would allow nested scripts which would be hella powerful). There would need to be four major syntaxes:

* `run_script: script_id` activates a script immediately.
* `stop_script: script_id` deactivates a running script.
* `add_script: [{script: script_id, trigger: trigger_id}]` would place new scripts in the NPCs script list exactly as if it was in the NPC definition, same syntax.
* `remove script: [{script: script_id, trigger: trigger_id}]` would remove scripts from the NPCs list. It is still important to specify the trigger in case you're running the same script more than once with different triggers.
* `pause_script` is another possible syntax that I don't see as being quite as important, but could have useful applications (for example, you could talk your way past some guards successfully, pausing their script to shoot you when you try to pass for 1 hour after which they presumably realize they've been duped.)

#### Triggers
A lot of the most powerful use of scripts would depend on having a number of good triggers for NPCs to run a particular script.  I've alluded to these above, but here are some that I think should be added:

* `enters_zone`, `faction_enters_zone`, `nonfaction_enters_zone`: Specify a square area of the current map by x,y coordinates (or perhaps specify a map zone by name/ID). If the target enters that zone within the NPC's line of sight, run the script.
* `leaves_zone`: As above, same criteria.
* `approach`: Specify a distance. If a creature meeting the criteria approaches within the specified distance and within line of sight, run the script.
* `interval`: Specify an interval, either a single value or an array of time units, eg `[ "5 m", "1 h" ]`.  The script will randomly run once per time interval (at a random speed if an array was specified).
* `death`, `ally_death`: The script will run if the character dies, or an ally within line of sight dies. This would finally allow us to set variables/effects to trigger on NPC death.

The following are some other possible triggers that aren't as immediately necessary but could have a great deal of useful application.

* `take_damage`, `ally_take_damage`: The script will run either if the script character takes damage, or an ally within line of sight takes damage
* `time`: The script runs at a particular time of day, eg 0600. I could use this to do things like have guards go on scheduled patrols.
* `hear_noise`: Triggers if the NPC hears any noise.
* `hear_phrase`: If the NPC hears a particular phrase (or list of phrases specified in an array), it runs the script. I could use this to set up scripted conversations between NPCs.

Triggers are a huge topic that could get quite fancy. Ideally we'd eventually have an entire trigger writing system, possibly defining them in JSON so that we could have built in conditionals, like if the player has a particular variable assigned, different things would happen... but for now, if we just had these things, there are all kinds of things we can do.

#### Script queue
(When I wrote this section I'd planned to add a lot more about looping scripts, but realized that can wait for a later iteration to add those. With time based triggers they might not even be necessary. So for now just ignore the bits about looping scripts)

There will need to be a script queue to determine behaviour if two scripts are activated simultaneously. The default behaviour should be:

* "run once" scripts are queued in the order they are received and process one at a time.
* "run once" scripts take priority over looping scripts. The looping scripts resumes once the run once script queue is complete
* only a single looping script can run at a time. A new looping script overrides any previous ones

There is also a `priority` defined in the script. If not specified, the priority is set to 10. If an incoming script has a higher priority, it goes above any scripts on the queue with a lower priority. A run once script or a new looping script will not interrupt a looping script with a higher priority. 

A good example of where priority might be used is if there is a script to open fire on non faction members that get too close to a valuable item. You might want your NPCs to drop what they're doing and go do that, rather than continue whatever they were up to.

**Comment by zachary-kaelan** [comment]
Chris Crawford has been talking about things like this for many years. [Here is a recent "course" explaining his vision.](http://www.erasmatazz.com/library/course-description-2018/index.html) Of particular interest might be his thoughts on [verbs](http://www.erasmatazz.com/library/course-description-2018/actual-verbs.html) and [personality models](http://www.erasmatazz.com/library/course-description-2018/a-more-artistic-example.html). It's meant to be self-contained, and a lot of it would be unnecessary for our purposes, but there's a lot of overlap with these plans for NPC AI.

**Comment by Xaleth** [comment]
I can see @Ramza13's effects_on_conditions working really well with this.
